presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -125,19 +127,68 @@ public BaseJdbcClient(JdbcConnectorId connectorId, BaseJdbcConfig config, String
         if (config.getConnectionPassword() != null) {
             connectionProperties.setProperty("password", config.getConnectionPassword());
         }
-        this.mapTableNames = config.isMapLowercaseTableNames();
+
+        // generate the map of lowercased schema names to JDBC schema names
+        schemaMappingCache = CacheBuilder.newBuilder().build(new CacheLoader<String, Optional<String>>()
+        {
+            @Override
+            public Optional<String> load(String key)
+                    throws Exception
+            {
+                for (String schemaName : getOriginalSchemas()) {
+                    if (schemaName.equals(key)) {
+                        return Optional.of(schemaName);
+                    }
+
+                    String schemaNameLower = schemaName.toLowerCase(ENGLISH);
+                    if (schemaNameLower.equals(key)) {
```

Comment:
Do we need to care about the case if mapping is equal to key? IMO it does not matter, but you would not need to store Optional in the cache and probably some `IF` blocks will also appear redundant.